### PR TITLE
Add user proxy docs. Make user proxy's default impl cancellable

### PR DESCRIPTION
**Pull request by peterychang**
Resolves: https://github.com/microsoft/autogen/issues/4332

### CODE REVIEW

File: python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py [hunk 1]
```diff
@@ -14,6 +15,13 @@
 InputFuncType = Union[SyncInputFunc, AsyncInputFunc]
 
 
+# TODO: ainput doesn't seem to play nicely with jupyter.
+async def cancellable_input(prompt: str, cancellation_token: Optional[CancellationToken]) -> Awaitable[str]:
```
Comment (jackgerrits) [inline python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py:19]: Rather than special casing input, we should ensure that any sync input function can be canceled if the token is cancelled  (line 80-82) of the original file

File: python/packages/autogen-core/pyproject.toml [hunk 2]
```diff
@@ -17,6 +17,7 @@ classifiers = [
 dependencies = [
     "openai>=1.3",
     "pillow",
+    "aioconsole",
```
Comment (jackgerrits) [inline python/packages/autogen-core/pyproject.toml:20]: Core shouldnt need this dependency

File: python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py [hunk 3]
```diff
@@ -14,6 +15,13 @@
...
+    task = asyncio.create_task(ainput(prompt))
+    cancellation_token.link_future(task)
+    return await task
+
+
 class UserProxyAgent(BaseChatAgent):
```
Comment (jackgerrits) [inline python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py:25]: Can you add a usage example to the class?

File: python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py [hunk 4]
```diff
@@ -51,7 +59,7 @@ def __init__(
     ) -> None:
         """Initialize the UserProxyAgent."""
         super().__init__(name=name, description=description)
-        self.input_func = input_func or input
+        self.input_func = input_func or cancellable_input
```
Comment (jackgerrits) [inline python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py:62]: Can you please add a test that showed a sync input function is cancellable

File: python/packages/autogen-core/docs/src/user-guide/agentchat-user-guide/tutorial/agents.ipynb
Comment (jackgerrits) [inline python/packages/autogen-core/docs/src/user-guide/agentchat-user-guide/tutorial/agents.ipynb:1]: Cancellation is too specific of a need to be in the user guide/tutorial for this agent and makes it confusing to read. This should just detail basics of using the agent and funnel through to API reference for advanced info.

Cancellation should be described in the reference

File: python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_magentic_one/_magentic_one_orchestrator.py
Comment (jackgerrits) [inline python/packages/autogen-agentchat/src/autogen_agentchat/teams/_group_chat/_magentic_one/_magentic_one_orchestrator.py:1]: Unrelated formatting changes

**Review by jackgerrits — CHANGES_REQUESTED** [review]

File: python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py (hunk 4 above)
Comment (peterychang) [reply python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py:62]: i dont think a sync input fn is cancellable

File: python/packages/autogen-core/pyproject.toml (hunk 2 above)
Comment (peterychang) [reply python/packages/autogen-core/pyproject.toml:20]: required for ainput

File: python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py [hunk 5]
```diff
@@ -40,6 +50,53 @@ class UserProxyAgent(BaseChatAgent):
...
+            from autogen_agentchat.agents import UserProxyAgent
+
+            agent = UserProxyAgent("user_proxy")
+            response = await asyncio.create_task(
+                agent.on_messages(
+                    [TextMessage(content="What is your name? ", source="user")],
+                    cancellation_token=token,
+                )
+            )
+            print(f"Your name is {response.chat_message.content}")
```
[the remark this answers is not in the document]
Comment (jackgerrits) [reply python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py:66]: Also top level awaits arent allowed in examples

File: python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py [hunk 6]
```diff
@@ -14,6 +15,15 @@
 InputFuncType = Union[SyncInputFunc, AsyncInputFunc]
 
 
+# TODO: ainput doesn't seem to play nicely with jupyter.
+#       No input window appears in this case.
+async def cancellable_input(prompt: str, cancellation_token: Optional[CancellationToken]) -> str:
+    task = asyncio.create_task(ainput(prompt))  # type: ignore
```
Comment (jackgerrits) [inline python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py:21]: cast the result to a string, then you can remove the rest of the type ignores

File: python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterychang) [reply python/packages/autogen-agentchat/src/autogen_agentchat/agents/_user_proxy_agent.py:66]: > Also top level awaits arent allowed in examples

Do I just need to wrap it into an async function def then?

File: python/packages/autogen-core/pyproject.toml (hunk 2 above)
Comment (jackgerrits) [reply python/packages/autogen-core/pyproject.toml:20]: That's needed in agentchat not core

File: python/packages/autogen-core/docs/src/user-guide/agentchat-user-guide/tutorial/agents.ipynb [hunk 7]
```diff
@@ -101,6 +101,39 @@
     "as well as a list of inner messages in the {py:attr}`~autogen_agentchat.base.Response.inner_messages` attribute,\n",
     "which stores the agent's \"thought process\" that led to the final response.\n",
     "\n",
+    "## User Proxy Agent\n",
+    "\n",
+    "{py:class}`~autogen_agentchat.agents.UserProxyAgent` is a built-in agent that\n",
+    provides one way for a user to intervene in the process. This agent will put the team in a temporary blocking state, and thus any exceptions or runtime failures while in the blocked state will result in a deadlock. It is strongly advised that this agent be coupled with a timeout mechanic and that all errors and exceptions emanating from it are handled.
```
Comment (peterychang) [inline python/packages/autogen-core/docs/src/user-guide/agentchat-user-guide/tutorial/agents.ipynb:107]:
```suggestion
    "provides one way for a user to intervene in the process. This agent will put the team in a temporary blocking state, and thus any exceptions or runtime failures while in the blocked state will result in a deadlock. It is strongly advised that this agent be coupled with a timeout mechanic and that all errors and exceptions emanating from it are handled."
```

**Review by jackgerrits — APPROVED** [review]
